// kons/dist/consola.js
import colors from 'picocolors';
import figures from 'figures';
import { f as formatMessage, c as createBadge, a as createIcon } from './utils-9ae856ad.js';

function createFormatter(iconOrType, iconColor, {
  target = console.log,
  textColor,
  isBadge,
  newline
} = {}) {
  const iconFormatter = isBadge ? createBadge : createIcon;
  const textColorFormatter = textColor ? colors[textColor] : colors[iconColor];
  return (...messages) => {
    const formattedMessages = formatMessage(messages);
    const icon = iconFormatter(iconOrType, iconColor);
    for (const message of formattedMessages) {
      target(`${icon} ${textColorFormatter(message)}${newline ? "\n" : ""}`);
    }
  };
}
const log = createFormatter(figures.pointer, "gray");
const info = createFormatter(figures.info, "blue", { target: console.info });
const warn = createFormatter(figures.warning, "yellow", { target: console.warn });
const success = createFormatter(figures.tick, "green");
const error = createFormatter("error", "red", { target: console.error, isBadge: true });

export { createFormatter, error, info, log, success, warn };
